crates/proto/src/udp/udp_client_stream.rs

```diff
@@ -24,7 +24,76 @@ use crate::udp::udp_stream::NextRandomUdpSocket;
 use crate::udp::{DnsUdpSocket, MAX_RECEIVE_BUFFER_SIZE};
 use crate::xfer::{DnsRequest, DnsRequestSender, DnsResponse, DnsResponseStream, SerialMessage};
 
-/// A UDP client stream of DNS binary packets
+/// A builder to create a UDP client stream.
+pub struct UdpClientStreamBuilder<P, MF = NoopMessageFinalizer> {
+    name_server: SocketAddr,
+    timeout: Option<Duration>,
+    signer: Option<Arc<MF>>,
+    bind_addr: Option<SocketAddr>,
+    provider: P,
+}
+
+impl<P, MF> UdpClientStreamBuilder<P, MF>
+where
+    MF: MessageFinalizer,
+{
+    /// Construct a new UDP client stream builder.
+    ///
+    /// # Arguments
+    ///
+    /// * `name_server` - the IP and Port of the DNS server to connect to
+    /// * `provider` - async runtime provider, for I/O and timers
+    pub fn new(name_server: SocketAddr, provider: P) -> Self {
```

Comment:
I'd suggest moving this into `UdpClientStream::builder()`.

https://matklad.github.io/2020/08/12/who-builds-the-builder.html